### PR TITLE
feat: stats command

### PR DESCRIPTION
**Pull request by andriyl**

### CODE REVIEW

**Comment by adamaltman** [comment]
Do you have a screenshot with the output? If so, add it to the PR description... I wonder if all of those colors will be readable in the terminal.

**Comment by andriyl** [comment]
> Do you have a screenshot with the output? If so, add it to the PR description... I wonder if all of those colors will be readable in the terminal.

@adamaltman 
Please have a look at how it looks right now:

image

I'm really curious about how to make the output better.

**Comment by adamaltman** [comment]
How can path items be more than operations?

**Comment by adamaltman** [comment]
Nevermind, a path item can be empty (without operations defined) — unusual and revealed by this stats command. 👍 

The order looks strange. What if we order it like this:

References
External Documents (is this 
Schema (Can you add this?)
Parameters (can you add this?)
Links
Path Items
Operations
Tags

**Comment by andriyl** [comment]
@adamaltman
Lists of stats are updated with the right order:
image
Also, the counting logic was improved as well.

**Comment by adamaltman** [comment]
Our other commands, like lint, also show how long it took to run. And they have other display formats. Is there a JSON display format available here too?

File: src/cli/stats-docs.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    PathMap: {
+      leave() {
+        statsCount.parameters.total = statsCount.parameters.items.size;
+        statsCount.refs.total = statsCount.refs.items.size;
+        statsCount.links.total = statsCount.links.items.size;
+      },
+      PathItem: {
+        leave() { statsCount.pathItems.total++; },
+        Operation: { leave() { statsCount.operations.total++; }},
+        Parameter: { leave(parameter: Oas2Parameter | Oas3Parameter) {
```
Comment (adamaltman) [inline src/cli/stats-docs.ts:85]: Parameter can be defined on the PathItem and on the Operation. I’m not clear if this catches both of those?

Comment (RomanHotsiy) [reply src/cli/stats-docs.ts:85]: It catches both.

**Comment by adamaltman** [comment]
🎉 

Can you please share the output for both `stylish` and `json` so that I can review?

**Comment by andriyl** [comment]
> 🎉
> 
> Can you please share the output for both `stylish` and `json` so that I can review?
yes, sure.

stylish:

image

json:

image

File: src/cli.ts [hunk 2]
```diff
@@ -12,14 +12,15 @@ import {
...
   promptUser,
   CircularJSONNotSupportedError,
+  getExecutionTime,
 } from './utils';
 import { formatProblems, OutputFormat } from './format/format';
 import { ResolveError, YamlParseError } from './resolve';
 import { loadConfig, Config, LintConfig } from './config/config';
 import { NormalizedProblem } from './walk';
 import { previewDocs } from './cli/preview-docs';
+import { handleStats } from './cli/stats-docs'
```
Comment (RomanHotsiy) [inline src/cli.ts:22]: rename file to just `stats`

File: src/cli/stats-docs.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+import { normalizeTypes } from '../types';
+import { Oas3Types } from '../types/oas3';
+import { Oas2Types } from '../types/oas2';
+import { WalkContext, walkDocument } from '../walk';
+import { normalizeVisitors, Oas2Visitor, Oas3Visitor } from '../visitors';
+import * as colors from 'colorette';
+import { getFallbackEntryPointsOrExit } from '../cli';
+import { Oas2Parameter } from '../typings/swagger';
+import { Oas3Parameter, OasRef } from '../typings/openapi';
+import { performance } from 'perf_hooks';
```
Comment (RomanHotsiy) [inline src/cli/stats-docs.ts:13]: move external packages imports to the top

File: src/cli/stats-docs.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+import { getExecutionTime } from '../utils';
+
+interface IStatsRow {
```
Comment (RomanHotsiy) [inline src/cli/stats-docs.ts:16]: No need to prefix interfaces with `I`. This is not common in TS

File: src/cli/stats-docs.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  metric: string;
+  total: number;
```
Comment (RomanHotsiy) [inline src/cli/stats-docs.ts:18]: What if some metric is not a number but e.g. a list of uniqute things or something more advanced?

Let's keep it simple for now but keep in mind we'll need other type of stats later.

File: src/cli/stats-docs.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+function printStatsStylish(statsCount: IStatsCount) {
+  for (const node in statsCount) {
+    const { metric, total, color } = statsCount[node as StatsName];
+    process.stderr.write(colors[color](`${metric}: ${total} \n`));
+  }
+}
+
+function printStatsJson(statsCount: IStatsCount) {
+  const json: any = {};
+  Object.keys(statsCount).forEach((key) => {
```
Comment (RomanHotsiy) [inline src/cli/stats-docs.ts:45]: let's use `for of` for consistency

File: src/cli/stats-docs.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      PathItem: {
+        leave() { statsCount.pathItems.total++; },
+        Operation: { leave() { statsCount.operations.total++; }},
+        Parameter: { leave(parameter: Oas2Parameter | Oas3Parameter) {
+          statsCount.parameters.items!.add(parameter.name)
+        }}
+      }
+    },
+    Components: {
+      Schema: { leave() { statsCount.schemas.total++; }}
```
Comment (RomanHotsiy) [inline src/cli/stats-docs.ts:119]: It won't catch OAS2 schemas: https://github.com/Redocly/openapi-cli/blob/master/src/types/oas2.ts#L16

Also, I would just use `NamedSchema` for OAS3 too. It should count all the top-level named schemas but can be defined on the top-level (no need to use nested visitor)

**Review by RomanHotsiy — COMMENTED** [review]
Looks really good 👍 

I would just make a Stats Visitor as a separate module somewhere instead of defining it inline.

I understand it relies on global stats var. Then maybe define it in global module scope instead of inline.

Also left some small comments.

**Review by RomanHotsiy — APPROVED** [review]

**Comment by RomanHotsiy** [comment]
Great job @andriyl 👏

**Pull request merged**